File: dotnet/KclLib.Tests/APITest.cs
namespace KclLib.Tests;

using KclLib.API;

[TestClass]
public class APITest
{
    static string parentDirectory = FindCsprojInParentDirectory(Environment.CurrentDirectory);

    [TestMethod]
    public void TestExecProgramAPI()
    {
        var api = new API();
        var execArgs = new ExecProgram_Args();
        var path = Path.Combine(parentDirectory, "test_data", "schema.k");
        execArgs.KFilenameList.Add(path);
        var result = api.ExecProgram(execArgs);
        Assert.AreEqual("app:\n  replicas: 2", result.YamlResult, result.ToString());
    }

    [TestMethod]
    public void TestExecProgramAPIFileNotFound()
    {
        var api = new API();
        var execArgs = new ExecProgram_Args();
        var path = Path.Combine(parentDirectory, "test_data", "file_not_found.k");
        execArgs.KFilenameList.Add(path);
        try
        {
            var result = api.ExecProgram(execArgs);
            Assert.Fail("No exception was thrown for non-existent file path");
        }
        catch (Exception ex)
        {
            Assert.AreEqual(true, ex.Message.Contains("Cannot find the kcl file"));
        }
    }

    [TestMethod]
    public void TestParseProgramAPI()
    {
        var path = Path.Combine(parentDirectory, "test_data", "schema.k");
        // Prepare arguments for parsing the KCL program
        var args = new ParseProgram_Args();
        args.Paths.Add(path);
        // Instantiate API and call parse_program method

        var result = new API().ParseProgram(args);

        // Assert the parsing results
        Assert.AreEqual(1, result.Paths.Count);
        Assert.AreEqual(0, result.Errors.Count);
    }

    [TestMethod]
    public void TestParseFileApi()
    {
        var path = Path.Combine(parentDirectory, "test_data", "schema.k");
        // Prepare arguments for parsing a single KCL file
        var args = new ParseFile_Args { Path = path };

        // Instantiate API and call parse_file method
        var result = new API().ParseFile(args);

        // Assert the parsing results
        Assert.AreEqual(0, result.Deps.Count);
        Assert.AreEqual(0, result.Errors.Count);
    }

    [TestMethod]
    public void TestOverrideFileAPI()
    {
        // Backup and restore test file for each test run
        string bakFile = Path.Combine(parentDirectory, "test_data", "override_file", "main.bak");
        string testFile = Path.Combine(parentDirectory, "test_data", "override_file", "main.k");
        File.WriteAllText(testFile, File.ReadAllText(bakFile));

        // Prepare arguments for overriding the KCL file
        var args = new OverrideFile_Args
        {
            File = testFile,
        };
        args.Specs.Add("b.a=2");
        // Instantiate API and call override_file method

        var result = new API().OverrideFile(args);

        // Assert the outcomes of the override operation
        Assert.AreEqual(0, result.ParseErrors.Count);
        Assert.AreEqual(true, result.Result);
    }

    [TestMethod]
    public void TestFormatPathAPI()
    {
        var api = new API();
        var args = new FormatPath_Args();
        var path = Path.Combine(parentDirectory, "test_data", "format_path", "test.k");
        args.Path = path;
        var result = api.FormatPath(args);
    }

    [TestMethod]
    public void TestFormatCodeAPI()
    {
        string sourceCode = "schema Person:\n" + "    name:   str\n" + "    age:    int\n" + "    check:\n"
                + "        0 <   age <   120\n";
        string expectedFormattedCode = "schema Person:\n" + "    name: str\n" + "    age: int\n\n" + "    check:\n"
                + "        0 < age < 120\n\n";
        var api = new API();
        var args = new FormatCode_Args();
        args.Source = sourceCode;
        var result = api.FormatCode(args);
        Assert.AreEqual(expectedFormattedCode, result.Formatted.ToStringUtf8(), result.ToString());
    }

    [TestMethod]
    public void TestGetSchemaTypeAPI()
    {
        var path = Path.Combine(parentDirectory, "test_data", "schema.k");
        var execArgs = new ExecProgram_Args();
        execArgs.KFilenameList.Add(path);
        var args = new GetSchemaTypeMapping_Args();
        args.ExecArgs = execArgs;
        var result = new API().GetSchemaTypeMapping(args);
        Assert.AreEqual("int", result.SchemaTypeMapping["app"].Properties["replicas"].Type, result.ToString());
    }

    [TestMethod]
    public void TestListOptionsAPI()
    {
        var path = Path.Combine(parentDirectory, "test_data", "option", "main.k");
        var args = new ParseProgram_Args();
        args.Paths.Add(path);
        var result = new API().ListOptions(args);
        Assert.AreEqual("key1", result.Options[0].Name);
        Assert.AreEqual("key2", result.Options[1].Name);
        Assert.AreEqual("metadata-key", result.Options[2].Name);
    }

    [TestMethod]
    public void TestListVariablesAPI()
    {
        var api = new API();
        var args = new ListVariables_Args();
        var path = Path.Combine(parentDirectory, "test_data", "schema.k");
        args.Files.Add(path);
        var result = api.ListVariables(args);
        Assert.AreEqual("AppConfig {replicas = 2}", result.Variables["app"].Variables[0].Value, result.ToString());
    }

    [TestMethod]
    public void TestLoadPackagesAPI()
    {
        var path = Path.Combine(parentDirectory, "test_data", "schema.k");
        var args = new LoadPackage_Args();
        args.ResolveAst = true;
        args.ParseArgs = new ParseProgram_Args();
        args.ParseArgs.Paths.Add(path);
        var result = new API().LoadPackage(args);
        var firstSymbol = result.Symbols.Values.FirstOrDefault();
        Assert.AreEqual(true, firstSymbol != null);
    }

    [TestMethod]
    public void TestLintPathAPI()
    {
        var path = Path.Combine(parentDirectory, "test_data", "lint_path", "test-lint.k");
        var args = new LintPath_Args();
        args.Paths.Add(path);
        var result = new API().LintPath(args);
        bool foundWarning = result.Results.Any(warning => warning.Contains("Module 'math' imported but unused"));
        Assert.AreEqual(true, foundWarning, result.ToString());
    }

    [TestMethod]
    public void TestValidateCodeAPI()
    {
        // Define the code schema and data
        string code = @"
schema Person:
    name: str
    age: int
    check:
        0 < age < 120
";
        string data = "{\"name\": \"Alice\", \"age\": 10}";

        // Prepare arguments for validating the code
        var args = new ValidateCode_Args
        {
            Code = code,
            Data = data,
            Format = "json"
        };

        // Instantiate API and call validate_code method

        var result = new API().ValidateCode(args);

        // Assert the validation results
        Assert.AreEqual(true, result.Success);
        Assert.AreEqual(string.Empty, result.ErrMessage);
    }

    [TestMethod]
    public void TestRenameAPI()
    {
        var root = Path.Combine(parentDirectory, "test_data", "rename");
        var renameFilePath = Path.Combine(parentDirectory, "test_data", "rename", "main.k");
        var renameBakFilePath = Path.Combine(parentDirectory, "test_data", "rename", "main.bak");
        File.WriteAllText(renameFilePath, File.ReadAllText(renameBakFilePath));
        var args = new Rename_Args
        {
            PackageRoot = root,
            SymbolPath = "a",
            NewName = "a2"
        };
        args.FilePaths.Add(renameFilePath);

        var result = new API().Rename(args);
        Assert.AreEqual(true, result.ChangedFiles.First().Contains("main.k"));
    }

    [TestMethod]
    public void TestRenameCodeAPI()
    {
        var args = new RenameCode_Args
        {
            PackageRoot = "/mock/path",
            SymbolPath = "a",
            SourceCodes = { { "/mock/path/main.k", "a = 1\nb = a" } },
            NewName = "a2"
        };

        var result = new API().RenameCode(args);
        Assert.AreEqual("a2 = 1\nb = a2", result.ChangedCodes["/mock/path/main.k"]);
    }

    [TestMethod]
    public void TestTestingAPI()
    {
        var pkg = Path.Combine(parentDirectory, "test_data", "testing");
        var args = new Test_Args();
        args.PkgList.Add(pkg + "/...");

        var result = new API().Test(args);
        Assert.AreEqual(2, result.Info.Count);
    }

    [TestMethod]
    public void TestLoadSettingsFilesAPI()
    {
        var workDir = Path.Combine(parentDirectory, "test_data");
        var settingsFile = Path.Combine(workDir, "settings", "kcl.yaml");
        var args = new LoadSettingsFiles_Args
        {
            WorkDir = workDir,
        };
        args.Files.Add(settingsFile);

        var result = new API().LoadSettingsFiles(args);
        Assert.AreEqual(0, result.KclCliConfigs.Files.Count);
        Assert.AreEqual(true, result.KclCliConfigs.StrictRangeCheck);
        Assert.AreEqual(true, result.KclOptions.Any(o => o.Key == "key" && o.Value == "\"value\""));
    }

    [TestMethod]
    public void TestUpdateDependenciesAPI()
    {
        var manifestPath = Path.Combine(parentDirectory, "test_data", "update_dependencies");
        // Prepare arguments for updating dependencies.
        var args = new UpdateDependencies_Args { ManifestPath = manifestPath };
        // Instantiate API and call update_dependencies method.

        var result = new API().UpdateDependencies(args);
        // Collect package names.
        var pkgNames = result.ExternalPkgs.Select(pkg => pkg.PkgName).ToList();
        // Assertions.
        Assert.AreEqual(2, pkgNames.Count);
    }

    [TestMethod]
    public void TestExecAPIWithExternalDependencies()
    {
        var manifestPath = Path.Combine(parentDirectory, "test_data", "update_dependencies");
        var testFile = Path.Combine(manifestPath, "main.k");
        // First, update dependencies.
        var updateArgs = new UpdateDependencies_Args { ManifestPath = manifestPath };

        var depResult = new API().UpdateDependencies(updateArgs);
        // Prepare arguments for executing the program with external dependencies.
        var execArgs = new ExecProgram_Args();
        execArgs.KFilenameList.Add(testFile);
        execArgs.ExternalPkgs.AddRange(depResult.ExternalPkgs);
        // Execute the program and assert the result.
        var execResult = new API().ExecProgram(execArgs);
        Assert.AreEqual("a: Hello World!", execResult.YamlResult);
    }

    [TestMethod]
    public void TestGetVersion()
    {
        var result = new API().GetVersion(new GetVersion_Args());
        Assert.AreEqual(true, result.VersionInfo.Contains("Version"), result.ToString());
        Assert.AreEqual(true, result.VersionInfo.Contains("GitCommit"), result.ToString());
    }

    static string FindCsprojInParentDirectory(string directory)
    {
        string parentDirectory = Directory.GetParent(directory).FullName;
        string csprojFilePath = Directory.GetFiles(parentDirectory, "*.csproj").FirstOrDefault();

        if (csprojFilePath != null)
        {
            return parentDirectory;
        }
        else if (parentDirectory == directory)
        {
            return null;
        }
        else
        {
            return FindCsprojInParentDirectory(parentDirectory);
        }
    }
}

File: python/tests/api_test.py
TEST_FILE = "./tests/test_data/schema.k"


def test_exec_api():
    """Execute KCL file with arguments and return the JSON/YAML result."""
    import kcl_lib.api as api

    # Call the `exec_program` method with appropriate arguments
    args = api.ExecProgram_Args(k_filename_list=[TEST_FILE])
    # Usage
    api = api.API()
    result = api.exec_program(args)
    assert result.yaml_result == "app:\n  replicas: 2"


def test_exec_api_failed():
    """Execute KCL file with arguments and return the JSON/YAML result."""
    import kcl_lib.api as api

    try:
        # Call the `exec_program` method with appropriate arguments
        args = api.ExecProgram_Args(k_filename_list=["file_not_found"])
        # Usage
        api = api.API()
        result = api.exec_program(args)
        assert False
    except Exception as err:
        assert "Cannot find the kcl file" in str(err)


def test_parse_program_api():
    """Parse KCL program with entry files and return the AST JSON string."""
    import kcl_lib.api as api

    args = api.ParseProgram_Args(paths=[TEST_FILE])
    api = api.API()
    result = api.parse_program(args)
    assert len(result.paths) == 1
    assert len(result.errors) == 0


def test_parse_file_api():
    """Parse KCL single file to Module AST JSON string with import dependencies and parse errors."""
    import kcl_lib.api as api

    args = api.ParseFile_Args(path=TEST_FILE)
    api = api.API()
    result = api.parse_file(args)
    assert len(result.deps) == 0
    assert len(result.errors) == 0


def test_load_package_api():
    """load_package provides users with the ability to parse KCL program and semantic model
    information including symbols, types, definitions, etc.
    """
    import kcl_lib.api as api

    # Call the `load_package` method with appropriate arguments
    args = api.LoadPackage_Args(
        parse_args=api.ParseProgram_Args(paths=[TEST_FILE]), resolve_ast=True
    )
    # Usage
    api = api.API()
    result = api.load_package(args)
    symbols = list(result.symbols.values())
    assert len([s for s in symbols if s.ty.schema_name == "AppConfig"]) > 0


def test_list_variables_api():
    """list_variables provides users with the ability to parse KCL program and get all variables by specs."""
    import kcl_lib.api as api

    # Call the `list_variable` method with appropriate arguments
    args = api.ListVariables_Args(files=[TEST_FILE])
    # Usage
    api = api.API()
    result = api.list_variables(args)
    assert (
        result.variables["app"].variables[0].value == "AppConfig {\n    replicas: 2\n}"
    )


def test_list_options_api():
    """list_options provides users with the ability to parse KCL program and get all option information."""
    import kcl_lib.api as api

    args = api.ParseProgram_Args(paths=["./tests/test_data/option/main.k"])
    api = api.API()
    result = api.list_options(args)
    assert len(result.options) == 3
    assert result.options[0].name == "key1"
    assert result.options[1].name == "key2"
    assert result.options[2].name == "metadata-key"


def test_get_schema_type_api():
    """Get schema type mapping defined in the program."""
    import kcl_lib.api as api

    exec_args = api.ExecProgram_Args(k_filename_list=[TEST_FILE])
    # Call the `get_schema_type_mapping` method with appropriate arguments
    args = api.GetSchemaTypeMapping_Args(exec_args=exec_args)
    # Usage
    api = api.API()
    result = api.get_schema_type_mapping(args)
    assert result.schema_type_mapping["app"].properties["replicas"].type == "int"


def test_override_file_api():
    """Override KCL file with arguments. See https://www.kcl-lang.io/docs/user_docs/guides/automation
    for more override spec guide.
    """
    import kcl_lib.api as api
    import pathlib

    bak_file = "./tests/test_data/override_file/main.bak"
    test_file = "./tests/test_data/override_file/main.k"

    pathlib.Path(test_file).write_text(pathlib.Path(bak_file).read_text())

    args = api.OverrideFile_Args(
        file=test_file,
        specs=["b.a=2"],
    )
    api = api.API()
    result = api.override_file(args)
    assert len(result.parse_errors) == 0
    assert result.result == True
    assert (
        pathlib.Path(test_file).read_text()
        == """\
a = 1
b = {
    "a": 2
    "b": 2
}
"""
    )


def test_format_code_api():
    """Format the code source."""
    import kcl_lib.api as api

    source_code = """\
schema Person:
    name:   str
    age:    int

    check:
        0 <   age <   120
"""

    args = api.FormatCode_Args(source=source_code)

    api_instance = api.API()
    result = api_instance.format_code(args)

    assert (
        result.formatted.decode()
        == """\
schema Person:
    name: str
    age: int

    check:
        0 < age < 120

"""
    )


def test_format_path_api():
    """Format KCL file or directory path contains KCL files and returns the changed file paths."""
    import kcl_lib.api as api

    TEST_PATH = "./tests/test_data/format_path/test.k"

    args = api.FormatPath_Args(path=TEST_PATH)

    api_instance = api.API()
    result = api_instance.format_path(args)

    assert result.changed_paths == []


def test_lint_path_api():
    """Lint files and return error messages including errors and warnings."""
    import kcl_lib.api as api

    TEST_PATH = "./tests/test_data/lint_path/test-lint.k"

    args = api.LintPath_Args(paths=[TEST_PATH])

    api_instance = api.API()
    result = api_instance.lint_path(args)

    assert "Module 'math' imported but unused" in result.results


def test_validate_code_api():
    """Validate code using schema and JSON/YAML data strings."""
    import kcl_lib.api as api

    code = """\
schema Person:
    name: str
    age: int

    check:
        0 < age < 120
"""
    data = '{"name": "Alice", "age": 10}'

    args = api.ValidateCode_Args(code=code, data=data, format="json")

    api_instance = api.API()
    result = api_instance.validate_code(args)

    assert result.success == True
    assert result.err_message == ""


def test_rename_api():
    """Rename all the occurrences of the target symbol in the files. This API will rewrite files if they contain symbols to be renamed.
    Return the file paths that got changed.
    """
    import kcl_lib.api as api
    import pathlib

    pathlib.Path("./tests/test_data/rename/main.k").write_text(
        pathlib.Path("./tests/test_data/rename/main.bak").read_text()
    )

    args = api.Rename_Args(
        package_root="./tests/test_data/rename",
        symbol_path="a",
        file_paths=["./tests/test_data/rename/main.k"],
        new_name="a2",
    )

    api_instance = api.API()
    result = api_instance.rename(args)

    assert "tests/test_data/rename/main.k" in result.changed_files[0]


def test_rename_code_api():
    """Rename all the occurrences of the target symbol and return the modified code if any code has been changed.
    This API won't rewrite files but return the changed code.
    """
    import kcl_lib.api as api

    args = api.RenameCode_Args(
        package_root="/mock/path",
        symbol_path="a",
        source_codes={"/mock/path/main.k": "a = 1\nb = a"},
        new_name="a2",
    )

    api_instance = api.API()
    result = api_instance.rename_code(args)

    assert result.changed_codes["/mock/path/main.k"] == "a2 = 1\nb = a2"


def test_testing_api():
    """Test KCL packages with test arguments."""
    import kcl_lib.api as api

    args = api.Test_Args(
        pkg_list=["./tests/test_data/testing/..."],
    )

    api_instance = api.API()
    result = api_instance.test(args)

    assert len(result.info) == 2


def test_load_settings_files_api():
    """Load the setting file config defined in `kcl.yaml`"""
    import kcl_lib.api as api

    args = api.LoadSettingsFiles_Args(
        work_dir="./tests/test_data", files=["./tests/test_data/settings/kcl.yaml"]
    )

    api_instance = api.API()
    result = api_instance.load_settings_files(args)

    assert result.kcl_cli_configs.files == []
    assert result.kcl_cli_configs.strict_range_check == True
    assert (
        result.kcl_options[0].key == "key" and result.kcl_options[0].value == '"value"'
    )


def test_update_dependencies_api():
    """Download and update dependencies defined in the `kcl.mod` file and return the
    external package name and location list.
    """
    import kcl_lib.api as api

    args = api.UpdateDependencies_Args(
        manifest_path="./tests/test_data/update_dependencies"
    )

    api_instance = api.API()
    result = api_instance.update_dependencies(args)
    pkg_names = [pkg.pkg_name for pkg in result.external_pkgs]

    assert len(pkg_names) == 2
    assert "helloworld" in pkg_names
    assert "flask" in pkg_names


def test_exec_api_with_external_dependencies():
    import kcl_lib.api as api

    args = api.UpdateDependencies_Args(
        manifest_path="./tests/test_data/update_dependencies"
    )

    api_instance = api.API()
    result = api_instance.update_dependencies(args)
    exec_args = api.ExecProgram_Args(
        k_filename_list=["./tests/test_data/update_dependencies/main.k"],
        external_pkgs=result.external_pkgs,
    )
    result = api_instance.exec_program(exec_args)
    assert result.yaml_result == "a: Hello World!"


def test_get_version_api():
    """Return the KCL service version information."""
    import kcl_lib.api as api

    api_instance = api.API()
    result = api_instance.get_version()
    assert "Version" in str(result) and "GitCommit" in str(result)
